internal/provider/resource_virtual_machine_group.go

```diff
@@ -0,0 +1,176 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/krystal/go-katapult/pkg/katapult"
+)
+
+func resourceVirtualMachineGroup() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceVirtualMachineGroupCreate,
+		ReadContext:   resourceVirtualMachineGroupRead,
+		UpdateContext: resourceVirtualMachineGroupUpdate,
+		DeleteContext: resourceVirtualMachineGroupDelete,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"segregate": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+		},
+	}
+}
+
+func resourceVirtualMachineGroupCreate(
+	ctx context.Context,
+	d *schema.ResourceData,
+	meta interface{},
+) diag.Diagnostics {
+	m := meta.(*Meta)
+
+	args := &katapult.VirtualMachineGroupCreateArguments{
+		Name: d.Get("name").(string),
+	}
+
+	if segregate := d.Get("segregate").(bool); segregate {
+		args.Segregate = &segregate
+	}
+
+	group, _, err := m.Client.VirtualMachineGroups.Create(
+		ctx, m.OrganizationRef(), args,
+	)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(group.ID)
```

Comment:
I don't think it matters too much here, but convention in various providers is to call `d.SetId()` as the last thing before the return. From what I've understood, it's mostly cause calling `d.SetId()` is what solidifies the resource into the TF state, so only do that once you've pulled out all the information you need successfully.